thread_sync.c

```diff
@@ -836,18 +836,48 @@ queue_closed_result(VALUE self, struct rb_queue *q)
  *
  */
 
+static ID id_each;
+static ID id_each_entry;
+
+static VALUE
+push_to_queue(RB_BLOCK_CALL_FUNC_ARGLIST(n, que))
+{
+    rb_ary_push(que, n);
+    return Qnil;
+}
+
 /*
  * Document-method: Queue::new
  *
- * Creates a new queue instance.
+ * Creates a new queue instance, optionally using the contents of an Enumerable
+ * for its initial state.
+ *
+ *  Example:
+ *
+ *    	q = Queue.new
+ *    	q = Queue.new([a, b, c])
+ *    	q = Queue.new(items)
  */
 
 static VALUE
-rb_queue_initialize(VALUE self)
+rb_queue_initialize(int argc, VALUE *argv, VALUE self)
 {
+    VALUE initial;
+    ID id_enum_method;
     struct rb_queue *q = queue_ptr(self);
     RB_OBJ_WRITE(self, &q->que, ary_buf_new());
     list_head_init(queue_waitq(q));
+    rb_scan_args(argc, argv, "01", &initial);
+    if (argc == 1) {
+        if (rb_respond_to(initial, id_each_entry)) {
+            id_enum_method = id_each_entry;
+        } else if (rb_respond_to(initial, id_each)) {
+            id_enum_method = id_each;
+        } else {
+            rb_raise(rb_eArgError, "value must be enumerable");
+        }
+        rb_block_call(initial, id_enum_method, 0, 0, push_to_queue, q->que);
```

Comment:
`rb_check_block_call` checks if the method is defined then calls it at once.

```suggestion
        if (rb_check_block_call(initial, id_each_entry, 0, 0, push_to_queue, q->que) == Qundef &&
            rb_check_block_call(initial, id_each, 0, 0, push_to_queue, q->que) == Qundef) {
            rb_raise(rb_eArgError, "value must be enumerable");
        }
```